src/lib.rs

```diff
@@ -27,38 +34,100 @@
 //! assert_eq!(weak.count(), 0);
 //! ```
 
+use notify::NotifySender;
 use std::fmt::{self, Display, Formatter};
 use std::sync::atomic::{AtomicUsize, Ordering};
 use std::sync::Arc;
 
+mod notify;
+
+pub use notify::{NotifyError, NotifyHandle, NotifyTimeoutError};
+
 /// Essentially an AtomicUsize that is clonable and whose count is based
-/// on the number of copies. The count is automatically updated on Drop.
+/// on the number of copies (and their size). The count is automatically updated on Drop.
+///
+/// If you want a weak reference to the counter that doesn't affect the count, see:
+/// [`WeakCounter`].
 #[derive(Debug)]
 pub struct Counter {
     counter: Arc<AtomicUsize>,
+    notify: Vec<NotifySender>,
     size: usize,
 }
 
-/// A 'weak' Counter that does not affect the count.
+/// A 'weak' [`Counter`] that does not affect the count.
 #[derive(Clone, Debug)]
-pub struct WeakCounter(Arc<AtomicUsize>);
+pub struct WeakCounter {
+    counter: Arc<AtomicUsize>,
+    notify: Vec<NotifySender>,
+}
 
-impl Counter {
-    pub fn new() -> Counter {
-        Counter::new_with_size(1)
+/// A builder for the [`Counter`].
+pub struct CounterBuilder {
+    counter: Arc<AtomicUsize>,
+    size: usize,
+    notify: Vec<NotifySender>,
+}
+
+impl CounterBuilder {
+    /// Change the specified size of the new [`Counter`]. This counter will add
+    /// `size` to the count, and will remove `size` from the count
+    /// when dropped.
+    pub fn size(mut self, v: usize) -> Self {
+        self.size = v;
+        self
     }
 
-    pub fn new_with_size(size: usize) -> Counter {
+    /// Create a [`NotifyHandle`] with a link to the count of this object. This [`NotifyHandle`] will
+    /// be notified when the value of this count changes.
+    ///
+    /// [`NotifyHandle`]s cannot be associated after creation, since all linked
+    /// [`Counter`] / [`WeakCounter`]s cannot be accounted for.
```

Comment:
not sure I follow this comment. what do you mean they can't be accounted for? if the notify were stored on the thing that's `Arc`'d, then there would be no issue right?

Reply:
The way I implemented it, the `NotifySender`s are references that are cloned from `Counter` to `Counter` as they get created, so in order to create a new `NotifyHandle` that is properly updated whenever the counter is changed, it would need to be given to all created `Counter` and `WeakCounter`s.

The only time we can guarantee that is during the initial creation of `WeakCounter` / `Counter`^.

Lemme know if that makes sense!

Reply:
Yeah makes sense, I think I was imagining that they all would share they could all share the same `Arc<Mutex<Vec<NotifyHandle>>` if you wanted to be able to update them after cloning, right?

Reply:
Yup true^